File: typegraph/python/typegraph/runtimes/wasmedge.py
# Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
# SPDX-License-Identifier: MPL-2.0
from dataclasses import dataclass
from typing import List, Optional

from typegraph import t
from typegraph.gen.exports.runtimes import (
    BaseMaterializer,
    Effect,
    EffectRead,
    MaterializerWasi,
)
from typegraph.gen.types import Err
from typegraph.runtimes.base import Materializer, Runtime
from typegraph.utils import get_file_hash
from typegraph.wit import runtimes, store


class WasmEdgeRuntime(Runtime):
    def __init__(self):
        super().__init__(runtimes.register_wasmedge_runtime(store))

    def wasi(
        self,
        inp: "t.struct",
        out: "t.typedef",
        *,
        func: str,
        wasm: str,
        effect: Optional[Effect] = None,
    ):
        effect = effect or EffectRead()
        artifact_hash = get_file_hash(wasm)
        wasm = f"file:{wasm}"

        mat_id = runtimes.from_wasi_module(
            store,
            BaseMaterializer(runtime=self.id.value, effect=effect),
            MaterializerWasi(module=wasm, func_name=func, artifact_hash=artifact_hash),
        )

        if isinstance(mat_id, Err):
            raise Exception(mat_id.value)

        return t.func(
            inp,
            out,
            WasiMat(id=mat_id.value, module=wasm, func_name=func, effect=effect),
        )


@dataclass
class WasiMat(Materializer):
    module: str
    func_name: str
    effect: List[str]

File: typegraph/deno/src/runtimes/wasmedge.ts
// Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

import * as t from "../types.ts";
import { runtimes } from "../wit.ts";
import { Effect } from "../../gen/exports/metatype-typegraph-runtimes.d.ts";
import { Materializer, Runtime } from "./mod.ts";

interface WasiMat extends Materializer {
  module: string;
  funcMame: string;
  effect: Effect;
}

export class WasmEdgeRuntime extends Runtime {
  constructor() {
    super(runtimes.registerWasmedgeRuntime());
  }

  wasi<
    P extends Record<string, t.Typedef> = Record<string, t.Typedef>,
    I extends t.Struct<P> = t.Struct<P>,
    O extends t.Typedef = t.Typedef,
  >(
    inp: I,
    out: O,
    { func, wasm, effect = { tag: "none" } }: {
      func: string;
      wasm: string;
      effect?: Effect;
    },
  ): t.Func<P, I, O, WasiMat> {
    const matId = runtimes.fromWasiModule(
      {
        runtime: this._id,
        effect,
      },
      {
        module: `file:${wasm}`,
        funcName: func,
      },
    );

    return t.func(inp, out, {
      _id: matId,
      effect,
      module: wasm,
      funcMame: func,
    });
  }
}
